linker: Introduce zephyr,memory-region compatible

carlocaione opened the pull request:
Introduce a new `zephyr,memory-region` compatible to be used when a new memory region must be created in the linker script from the devicetree node using the compatible.

Remove also the `LINKER_DT_REGION_FROM_NODE` macro and add a new `LINKER_DT_REGION` macro to cycle through all the compatible regions.

This is a preparatory work to be integrated with #41223 so that under one single compatible we have all the properties to create a new region in the linker script and set the MPU properties of the region using the info coming from the DT.

carlocaione committed:
linker: Introduce zephyr,memory-region compatible

Introduce a new "zephyr,memory-region" compatible to be used when a new
memory region must be created in the linker script from the devicetree
nodes using the compatible.

Remove also the LINKER_DT_REGION_FROM_NODE macro and add a new
LINKER_DT_REGIONS macro to cycle through all the compatible regions.

In the same PR modify the DTS files and the linker scripts.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

JordanYates commented:
AFAICT we've lost the `rw` or `rwx` attributes on the memory regions with this PR.
I suspect we'll need an additional property called `zephyr,memory-region-attr` or similar

carlocaione commented:
> AFAICT we've lost the `rw` or `rwx` attributes on the memory regions with this PR.
> I suspect we'll need an additional property called `zephyr,memory-region-attr` or similar

Yes, but do we _really_ need those? AFAICT those are attributes are optional and they will conflict with the MPU configuration parameter that I want to add later, so I would gladly leave that out if possible (how are used by the toolchain?)

JordanYates commented:
> Yes, but do we really need those? AFAICT those are attributes are optional and they will conflict with the MPU configuration parameter that I want to add later, so I would gladly leave that out if possible (how are used by the toolchain?)

Based on the description here, https://sourceware.org/binutils/docs/ld/MEMORY.html, it looks like they're only used as a fallback of where to place sections that are not explicitly placed. All of our sections should be explicitly placed AFAIK and the `FLASH` and `SRAM` regions still have their attributes, so I'm guessing it should be fine.

mbolivar-nordic commented on include/linker/devicetree_regions.h:
```diff
@@ -43,21 +45,16 @@
-	COND_CODE_1(DT_NODE_HAS_STATUS(node_id, okay), \
-		    (_REGION_DECLARE(node_id, attr)),  \
-		    ())
+#define LINKER_DT_REGION() \
```
I think it would be clearer to call this `LINKER_DT_REGIONS`, since it declares multiple regions, not just one region.

Nukersson replied:
```suggestion
#define LINKER_DT_REGION(attr) \
```

mbolivar-nordic reviewed (commented):
No objections from the devicetree side if this works for both you and @JordanYates . I'd like @tejlmand to take a look at the build system portion since I don't see support on the armclang side.

One minor naming nit.

mbolivar-nordic commented:
> No objections from the devicetree side if this works for both you and @JordanYates

I also agree with your comment here (https://github.com/zephyrproject-rtos/zephyr/pull/41223#issuecomment-1018270948) that @Nukersson should take a look at this before merging too; sorry for missing that at first.

Nukersson commented:
> > AFAICT we've lost the `rw` or `rwx` attributes on the memory regions with this PR.
> > I suspect we'll need an additional property called `zephyr,memory-region-attr` or similar
> 
> Yes, but do we _really_ need those? AFAICT those are attributes are optional and they will conflict with the MPU configuration parameter that I want to add later, so I would gladly leave that out if possible (how are used by the toolchain?)

@JordanYates @carlocaione I think, leaving attributes away breaks the interface. Can the attributes stay untouched? Is it technically difficult to implement?

Nukersson commented on soc/riscv/riscv-privilege/andes_v5/ae350/linker.ld:
```diff
@@ -87,9 +87,7 @@ MEMORY
-    /* Data & Instruction Tightly Coupled Memory */
-    LINKER_DT_REGION_FROM_NODE(DT_CHOSEN(zephyr_itcm), rw)
-    LINKER_DT_REGION_FROM_NODE(DT_CHOSEN(zephyr_dtcm), rw)
+    LINKER_DT_REGION()
```
```suggestion
    LINKER_DT_REGION(rw)
```

Nukersson commented on include/arch/arm/aarch32/cortex_a_r/scripts/linker.ld:
```diff
@@ -81,7 +81,7 @@ MEMORY
     FLASH    (rx) : ORIGIN = ROM_ADDR, LENGTH = ROM_SIZE
     SRAM     (wx) : ORIGIN = RAM_ADDR, LENGTH = RAM_SIZE
-    LINKER_DT_REGION_FROM_NODE(DT_CHOSEN(zephyr_ocm), rw)
+    LINKER_DT_REGION()
```
```suggestion
    LINKER_DT_REGION(rw)
```

Nukersson commented on include/arch/arm/aarch32/cortex_m/scripts/linker.ld:
```diff
@@ -81,22 +81,7 @@ MEMORY
-    LINKER_DT_REGION_FROM_NODE(DT_NODELABEL(sdram1), rw)
-    LINKER_DT_REGION_FROM_NODE(DT_NODELABEL(sdram2), rw)
-    LINKER_DT_REGION_FROM_NODE(DT_NODELABEL(backup_sram), rw)
+    LINKER_DT_REGION()
```
```suggestion
    LINKER_DT_REGION(rw)
```
1. Attributes are not used. 2. Just ignore rwx and use rw for all since not used in later implementation.

Nukersson commented on include/arch/riscv/common/linker.ld:
```diff
@@ -81,9 +81,7 @@ MEMORY
-    /* Data & Instruction Tightly Coupled Memory */
-    LINKER_DT_REGION_FROM_NODE(DT_CHOSEN(zephyr_itcm), rw)
-    LINKER_DT_REGION_FROM_NODE(DT_CHOSEN(zephyr_dtcm), rw)
+    LINKER_DT_REGION()
```
```suggestion
    LINKER_DT_REGION(rw)
```

Nukersson commented on include/linker/devicetree_regions.h:
```diff
@@ -43,21 +45,16 @@
-#define _REGION_DECLARE(node_id, attr)		    \
-	LINKER_DT_NODE_REGION_NAME(node_id)(attr) : \
-	ORIGIN = DT_REG_ADDR(node_id),		    \
+#define _REGION_DECLARE(node_id)	      \
+	LINKER_DT_NODE_REGION_NAME(node_id) : \
```
```suggestion
#define _REGION_DECLARE(node_id, attr)	      \
	LINKER_DT_NODE_REGION_NAME(node_id)(attr) : \
```

Nukersson commented on include/linker/devicetree_regions.h:
```diff
@@ -43,21 +45,16 @@
-		    (_REGION_DECLARE(node_id, attr)),  \
-		    ())
+#define LINKER_DT_REGION() \
+	DT_FOREACH_STATUS_OKAY(_DT_COMPATIBLE, _REGION_DECLARE)
```
```suggestion
	DT_FOREACH_STATUS_OKAY(_DT_COMPATIBLE, _REGION_DECLARE(attr))
```
I am not sure here, if it is correct suggestion, how to use attributes in this context.

KozhinovAlexander reviewed (changes requested):
@carlocaione Great work and respect for pushing this very important topic forward!

Please add  take a look on my review and changes and agree or disagree with proof. I hope, it is not a huge amount work to do for you 😃

JordanYates commented:
> The 2nd should accept an array (ideally) or proposed zephyr,memory-region-attr. But you're lucky here, because the attributes are same for all the regions and not used 😄. Thus yo can use one attribute only for sacke of simplicity. Later somebody else can add the array, if there would be a need.

That is not the case currently, the `ti-ccfg` region uses `rwx`

Nukersson commented:
I've proposed a change with comment, why rw can be done instead of rwx. Please take a look on it.

Nukersson commented:
@carlocaione Well, you do not have to implement attributes propagation to MPU. Just leave them where they were and put comment on the macro, that attributes are not used and will stay here for compatibility reasons. I think, there should be nothing thrown away.

carlocaione commented:
> @carlocaione Well, you do not have to implement attributes propagation to MPU. Especially you do not have to use them in your code. Just leave them as arguments of the macros as they are.

Ok, if I do not have to propagate the attributes to the MPU, I do not have to use them in the code, they are not used anywhere, why are we bothering to leave those in place?

> Please put a comment on the macro, that attributes are not used and will stay here for compatibility reasons. I think, there should be nothing thrown away.

Compatibility with what exactly?

Nukersson commented:
Obviously attributes are a dead code, I agree with you here. But we do not know for what the were originally introduced, since we do not have documentation. Therefore it is better to leave them at place, since there is no proof, they were introduced for some reason. Just do not delete old code or maybe not functional code, if you do not exactly know, where it can hit.

carlocaione commented:
> Obviously attributes are a dead code, I agree with you here. But we do not know for what they were originally introduced, since we do not have documentation.

The documentation is clear on what the attributes are for (see https://sourceware.org/binutils/docs/ld/MEMORY.html) and for what Zephyr is concerned we do not have an actual use for those.

> Therefore it is better to leave them at place, since there is no proof, they were introduced for some reason. Just do not delete old code or maybe not functional code, if you do not exactly know, where it can hit.

This is not possible. We have two ways here:
- We remove the attributes because there is no use for those and they will conflict with the MPU configuration
- We fully support the attributes because I do not want to introduce technical debt. That means that we have to support all the possible combinations (`r`, `rw`, `rwx`, etc...) by adding a new DT property and we have to find a way to make this co-exist with MPU configuration.

> Effectively, from my side I would give this PR my approve also with deleted attributes, cause it is a good working PR.

I'll add a comment explaining why the attributes are missing, thanks.

gmarull commented on include/linker/devicetree_regions.h:
```diff
@@ -7,6 +7,8 @@
  * Generate memory regions from devicetree nodes.
  */
 
+#define _DT_COMPATIBLE	zephyr_memory_region
```
nit: internal helper macros should all be put under

```
/** @cond INTERNAL_HIDDEN */
...
/** @endcond */
```

carlocaione replied:
In that file there is already an `INTERNAL_HIDDEN` section, so I'm just moving the `_DT_COMPATIBLE` in there.

gmarull reviewed (dismissed):
One minor documentation comment

carlocaione commented:
@tejlmand can you take a look?

tejlmand reviewed (changes requested):
Great work.

We need to be able to be able to fetch all `zephyr,memory-region` compatibles in same way here:
https://github.com/zephyrproject-rtos/zephyr/blob/a03deb8ef3c940bac125e35442ba00bd2e6f51bb/cmake/linker_script/arm/linker.cmake#L37-L50

so that code can use same principle.

Everything else looks good.
Let me know if you would like me to take a look at it.

carlocaione committed:
cmake: dts: Add dt_comp_path() cmake function

Add a new cmake extension function:

  dt_comp_path(<var> COMPATIBLE <compatible> [INDEX <idx>])

to get a list of paths for the nodes with the given <compatible>. This
is useful when we have to cycle through several nodes with the same
compatible.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

carlocaione committed:
cmake: linker: Automatically generates memory regions

As already done for the regular linker script, dinamically generates the
memory regions with the 'zephyr,memory-region' compatible also when
using the cmake linker generator.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

carlocaione commented:
@tejlmand I gave it a shot.

@mbolivar-nordic you want probably take a look at this again.

tejlmand commented on cmake/extensions.cmake:
```diff
@@ -2819,6 +2819,59 @@ function(dt_prop var)
+
+  foreach(arg ${req_single_args})
+    if(NOT DEFINED DT_COMP_${arg})
+      message(FATAL_ERROR "dt_comp(${ARGV0} ...) "
```
Function is called: `dt_comp_path()`

```suggestion
      message(FATAL_ERROR "dt_comp_path(${ARGV0} ...) "
```

carlocaione replied:
right

tejlmand commented on soc/riscv/riscv-privilege/andes_v5/ae350/linker.ld:
```diff
@@ -87,9 +87,7 @@ MEMORY
-    /* Data & Instruction Tightly Coupled Memory */
-    LINKER_DT_REGION_FROM_NODE(DT_CHOSEN(zephyr_itcm), rw)
-    LINKER_DT_REGION_FROM_NODE(DT_CHOSEN(zephyr_dtcm), rw)
+    LINKER_DT_REGIONS()
```
When testing with `arty_a7_arm_designstart_m3` this code results in `DTCM` and `ITCM` being defined on the same line like this:

```
MEMORY
    {
    FLASH (rx) : ORIGIN = (0x0 + 0x0), LENGTH = (32*1K - 0x0)
    SRAM (wx) : ORIGIN = 0x20000000, LENGTH = (32 * 1K)
    "ITCM" : ORIGIN = 0, LENGTH = 32768 "DTCM" : ORIGIN = 536870912, LENGTH = 32768
    IDT_LIST (wx) : ORIGIN = 0xFFFFF7FF, LENGTH = 2K
    }
```

it's not invalid code, but would be nice if each could have their own line.

This looks very nice when using `-DCONFIG_CMAKE_LINKER_GENERATOR=y` :wink:

tejlmand replied:
I agree, not easy at all and I can't think of a solution, just wanted to mention it.

tejlmand reviewed (commented):
Looks very good, thanks for putting this together.

A couple of very minor nits.

carlocaione commented on cmake/extensions.cmake:
```diff
@@ -2819,6 +2819,59 @@ function(dt_prop var)
+  cmake_parse_arguments(DT_COMP "" "${req_single_args};${single_args}" "" ${ARGN})
+
+  if(${ARGV0} IN_LIST req_single_args)
+    message(FATAL_ERROR "dt_comp(${ARGV0} ...) missing return parameter.")
```
yup

carlocaione commented on cmake/linker_script/arm/linker.cmake:
```diff
@@ -34,20 +34,11 @@ zephyr_linker_memory(NAME FLASH    FLAGS rx START ${FLASH_ADDR} SIZE ${FLASH_SIZ
-zephyr_linker_dts_memory(NODELABEL sdram2      FLAGS rw)
-zephyr_linker_dts_memory(NODELABEL backup_sram FLAGS rw)
+# Only use 'rw' as FLAGS. It's not used anyway.
+dt_comp_path(paths COMPATIBLE "zephyr,memory-region")
+foreach(path IN LISTS paths)
+  zephyr_linker_dts_memory(PATH ${path} FLAGS rw)
+endforeach()
```
;)

mbolivar-nordic commented on scripts/dts/gen_dts_cmake.py:
```diff
@@ -95,6 +96,7 @@ def main():
         for alias in node.aliases:
             cmake_props.append(f'"DT_ALIAS|{alias}" "{path}"')
 
+    comp_iter = defaultdict(list)
```
Naming nitpick here. `iter` is a built-in python function that returns an iterator over an iterable. So the name `comp_iter` suggests it is an iterator. But it's not -- it's a dict, which is not an iterator.

I suggest renaming this `compatible2paths` to match the general edtlib convention of naming a dict `<key>2<value>`

carlocaione replied:
yup, done.

mbolivar-nordic reviewed (approved):
I'm happy with the devicetree portions; @tejlmand should decide about the build system parts.

One nonblocking suggestion for the script.

Thanks for your persistence, @carlocaione.

tejlmand reviewed (approved):
Looks good.

Thanks for the effort.
